forms/session_form.go

```diff
@@ -0,0 +1,56 @@
+package forms
+
+import (
+	"google-scraper/helpers"
+	"google-scraper/models"
+
+	"github.com/beego/beego/v2/adapter/logs"
+	"github.com/beego/beego/v2/core/validation"
+)
+
+type SessionForm struct {
+	Email    string `form:"email" valid:"Email; Required"`
+	Password string `form:"password" valid:"Required;"`
+}
+
+var currentUser *models.User
+
+func (sessionForm *SessionForm) Valid(v *validation.Validation) {
+	errMessage := "Incorrect email or password"
+	logMessage := "Setting error on validation failed: "
+
+	user, err := models.FindUserByEmail(sessionForm.Email)
+	if err != nil {
+		err := v.SetError("Email", errMessage)
+		if err == nil {
+			logs.Error(logMessage, err)
+		}
+	} else {
+		err = helpers.CheckPasswordHash(sessionForm.Password, user.HashedPassword)
+		if err != nil {
+			err := v.SetError("Password", errMessage)
+			if err == nil {
+				logs.Error(logMessage, err)
+			}
+		} else {
+			currentUser = user
+		}
+	}
+}
+
+func (sessionForm *SessionForm) Authenticate() (*models.User, error) {
+	validation := validation.Validation{}
+	success, err := validation.Valid(sessionForm)
+
+	if err != nil {
+		logs.Error("Validation error:", err)
+	}
+
+	if !success {
+		for _, err := range validation.Errors {
+			return nil, err
```

Comment:
If there is more than one error, this will return on the first errors. Is that intended? Pae implemented differently by returning a slice of errors 👉 https://github.com/carryall/go-google-scraper-challenge/blob/2564bfdc83374763ec12ea6cea03dcfa3bcc4054/forms/session.go#L51-L55 

In terms of security, it's better to have a generic error message as done here "Incorrect email or password" to avoid user enumeration. In that case, there would be no need to loop over the errors at all. If there is any error, just return the generic error. 

You need to decide which way you want to go with.

Reply:
Thanks, I chose  to go with the generic message. I think it is good for security concern